### ISSUE TITLE
Screen tearing

### ISSUE THREAD

**Issue by Matthaiks** [narzoul/DDrawCompat#3]
DDrawCompat is causing screen tearing in some games. For instance, in the Starcraft: Brood War intro.
Sadly, GPU V-sync does nothing.

**Comment by narzoul** [comment]
I can't see any tearing in the Brood War intro myself. I checked how it is being rendered, and like the rest of the game it doesn't use double buffering, which means driver v-sync settings won't affect it. It writes directly to the primary surface at a rate of 15 fps.
Although it does attempt to synchronize with the vertical blank every 1-2 seconds, this can only prevent tearing if your monitor's refresh rate is an exact multiple of 15. Could you check what is the refresh rate when playing Brood War?

To be fair, in 256 color games DDrawCompat currently does have some delay in rendering frames (about 2ms as I measured on my system) which could potentially be reduced by storing palettized surfaces in system memory instead of video memory. This is something I plan to add in a future release.
This delay may introduce some tearing around the top of the screen, but there is a large enough black bar there (in Brood War intro) that I think it would be constrained to that region (and thus invisible in this case), unless your monitor is running at a very high refresh rate (120+ Hz?), or perhaps if the delay is larger on your system.

That said, v-sync in general can be a rather complex topic, so let's break down the general cases and issues too.

First of all, you probably won't have the same tearing in most games if you don't use DDrawCompat. The reason for this is that Windows Vista and above runs full-screen DirectDraw games in a fake full-screen mode ("maximized windowed mode") by default, which means that the display is actually still controlled by the desktop compositor, which always enforces v-sync (can't be overridden from drivers either). A nice side-effect is that you get automatic v-sync (and probably reduced tearing as a result) even if the game doesn't use double buffering or doesn't otherwise support v-sync.

DDrawCompat disables this windowed mode for several reasons: it's easier to implement GDI emulation without it, v-sync can actually be controlled, and performance can be significantly improved on Windows 8. There is an alternative fix for the Win 8 performance issues: the "NoGdiHWAcceleration" compatibility flag in Microsoft Application Compatibility Toolkit. But I haven't figured out a way to enable this flag from DDrawCompat yet, and until then I'm pretty much forced to stick to "real" full-screen mode at least for performance reasons.

As I already mentioned, games that don't use double (or triple) buffering will normally have tearing issues (in real full-screen mode), unless the game itself implements some kind of synchronized rendering (unlikely). In this case the only thing that can be done is to implement this kind of synchronization in DDrawCompat itself. This is possible (though not simple), but it should probably be a configurable option, as some people prefer to have higher frame rates in spite of tearing. I will consider this after configurability in general is implemented.

Games with double/triple buffering are simple enough from a v-sync perspective. In DirectDraw, these automatically get v-sync unless the game itself explicitly requests otherwise. In either case the decision can still be overridden by driver config. DDrawCompat already respects these settings, though again it would be possible to add a config option to override this from the wrapper as well.

To sum it up, this issue will probably remain open for a long time, but eventually I do plan to introduce proper v-sync support even in the single buffered cases.

**Comment by Matthaiks** [comment]
Thanks for the detailed explanation. 
According to my monitor, the whole game is running 640x480@60Hz (it's a 1920x1200 60Hz monitor).
BTW, dgVoodoo2 causes tearing in SC too. It has the "Force vSync" option and it helps with cinematics but slows down everything else.

**Comment by narzoul** [comment]
I've added a sort of forced v-sync for games that don't use page flipping for presentation. It would probably work for StarCraft too, except the game has since been updated and no longer seems to use DirectDraw. Anyway, I consider this issue fixed and will close it now. If you can still test an older StarCraft version and v-sync doesn't work as expected, feel free to reopen it.

**Comment by Matthaiks** [comment]
SC 1.16.1 is working fine now. Thanks!

BTW, experimental versions break Diablo 1.

**Comment by narzoul** [comment]
I've looked into Diablo 1. I assume you mean how the palette gets corrupted when you move the mouse near the bottom of the screen?

That appears to be a bug in the game itself. It doesn't respect the image boundaries of the primary surface correctly and writes to 2 rows over the bottom of the buffer, causing memory corruption. In this case it looks like it overwrites the color palette, most obviously as you move the cursor around the bottom edge of the screen.

This issue surfaced after I made a change to create palettized surfaces in system memory for efficiency reasons. Normally I guess DirectDraw would put them in video memory, which doesn't seem to be affected as much by this corruption. Maybe the locking mechanism doesn't even allow writing past the video memory buffer boundaries, or at least there is no palette stored there.

The obvious fix is the resize the system memory primary surface buffers and add at least 2 extra rows. A quick experiment shows that this fixes the palette corruption issue. Interestingly, if I don't add at least 2 rows, on Intel HD I get a completely black screen after the Blizzard logo movie.

I'll reopen this issue as a reminder until a final fix is committed, to make sure I don't forget about it.

**Comment by Matthaiks** [comment]
Sometimes Diablo 1 hangs at startup (black screen), sometimes it works but the screen flashes white while browsing the main menu and there's an application error on exit (atiumdva.dll).

**Comment by narzoul** [comment]
Yes, those are similar to the issues I had prior to the memory corruption fix. Please check the latest experimental release, it should all work fine now.

**Comment by Matthaiks** [comment]
It's working fine now, thanks!

**Comment by Matthaiks** [comment]
Carmageddon 95 works only in the main menu with experimental versions. Crash to desktop during race loading.

**Comment by narzoul** [comment]
Thanks for the report, it should be fixed now. Although I noticed there's some flashing leftover graphics at the top/bottom of the screen when going back to the menu from a race. I will look into that also.

**Comment by Matthaiks** [comment]
Thanks for fixing it!

That flashing is also present when using dgVoodoo2. cnc-ddraw is fine, though.

**Comment by narzoul** [comment]
Latest experimental release should fix the flashing menu problem.

**Comment by Matthaiks** [comment]
Fixed. Thanks!

**Comment by Matthaiks** [comment]
Another regression:
Win 95 versions of Command & Conquer and C&C: Red Alert refuse to start (black screen).

**Comment by narzoul** [comment]
Thanks again. I actually knew about the issue with Red Alert for a while now, but only just got around to fixing it. I tested the solution with C&C Gold also and it should work too now. Check the latest experimental release.

**Comment by Matthaiks** [comment]
Thanks for fixing it.

**Comment by narzoul** [comment]
Unfortunately I had to revert the fix for Diablo 1 because it's causing a BSOD with the Drakan Demo in windowed mode, at least on AMD drivers (works fine with Intel HD). I'll need to investigate this further and find a better solution.
